gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java

```diff
@@ -334,78 +327,136 @@ private static <T> ObjectConstructor<T> newDefaultImplementationConstructor(
      */
 
     if (Collection.class.isAssignableFrom(rawType)) {
-      if (SortedSet.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new TreeSet<>();
-          }
-        };
-      } else if (Set.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new LinkedHashSet<>();
-          }
-        };
-      } else if (Queue.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ArrayDeque<>();
-          }
-        };
-      } else {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ArrayList<>();
-          }
-        };
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newCollectionConstructor(rawType);
+      return constructor;
     }
 
     if (Map.class.isAssignableFrom(rawType)) {
-      if (ConcurrentNavigableMap.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ConcurrentSkipListMap<>();
-          }
-        };
-      } else if (ConcurrentMap.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new ConcurrentHashMap<>();
-          }
-        };
-      } else if (SortedMap.class.isAssignableFrom(rawType)) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new TreeMap<>();
-          }
-        };
-      } else if (type instanceof ParameterizedType
-          && !String.class.isAssignableFrom(
-              TypeToken.get(((ParameterizedType) type).getActualTypeArguments()[0]).getRawType())) {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new LinkedHashMap<>();
-          }
-        };
-      } else {
-        return new ObjectConstructor<T>() {
-          @Override
-          public T construct() {
-            return (T) new LinkedTreeMap<>();
-          }
-        };
-      }
+      @SuppressWarnings("unchecked")
+      ObjectConstructor<T> constructor = (ObjectConstructor<T>) newMapConstructor(type, rawType);
+      return constructor;
+    }
+
+    // Unsupported type; try other means of creating constructor
+    return null;
```

Comment:
In https://github.com/google/gson/issues/419#issuecomment-83705342 the concern was mentioned that the fallback will be `Unsafe` which might leave the Collection or Map implementation in a broken state because fields have not been initialized. And the built-in Collection and Map adapters might then encounter confusing exceptions when trying to add elements.

Not sure if we should throw an exception then in case of Collection or Map? Or we could maybe add an `allowUnsafe` parameter here (which is considered in addition to `GsonBuilder#disableJdkUnsafe`), which the built-in Collection and Map adapters can set to `false` because calling `add` and `put` will likely fail afterwards.